proposals/safe-fixed-size-buffers.md

```diff
@@ -0,0 +1,558 @@
+- Otherwise, the expression is classified as a value of type `System.ReadOnlySpan<S>`, where S is the element type of the fixed-size buffer.
+  The value can be used to access member's elements.
+
+### Definite assignment checking
+
+Fixed-size buffers are not subject to definite assignment-checking, and fixed-size buffer members are ignored for purposes of definite-assignment checking of struct type variables.
+
+When a fixed-size buffer member is static or the outermost containing struct variable of a fixed-size buffer member is a static variable, an instance variable of a class instance, or an array element, the elements of the fixed-size buffer are automatically initialized to their default values. In all other cases, the initial content of a fixed-size buffer is undefined.
+
+### Metadata
+
+#### Metadata emit and code generation
+
+For metadata encoding compiler will rely on recently added [```System.Runtime.CompilerServices.InlineArrayAttribute```](https://github.com/dotnet/runtime/issues/61135). 
+
+Fixed-size buffers like:
+``` C#
+public partial class C
+{
+    public int buffer1[10];
+    public readonly int buffer2[10];
+}
+```
+will be emitted as  fields of a specially decorated struct type.
+
+Equivalent C# code will be:
+
+``` C#
+public partial class C
+{
+    public Buffer10<int> buffer1;
+    public readonly Buffer10<int> buffer2;
+}
+
+[System.Runtime.CompilerServices.InlineArray(10)]
+public struct Buffer10<T>
+{
+    private T _element0;
+
+    [UnscopedRef]
+    public System.Span<T> AsSpan()
+    {
+        return System.Runtime.InteropServices.MemoryMarshal.CreateSpan(ref _element0, 10);
+    }
+
+    [UnscopedRef]
+    public readonly System.ReadOnlySpan<T> AsReadOnlySpan()
+    {
+        // Note, an attempt to compile reports error CS1605: Cannot use '_element0' as a ref or out value because it is read-only
+        return System.Runtime.InteropServices.MemoryMarshal.CreateReadOnlySpan(ref _element0, 10);  
+    }
+}
+```
+
+The actual naming conventions for the type and its members are TBD. The framework will likely include a set of predefined "buffer" types that cover
+a limited set of buffer sizes. When a predefined type doesn't exist, compiler will synthesize it in the module being built. Names of the generated types
+will be "speakable" in order to support consumption from other languages. 
+
+A code generated for an access like: 
+``` C#
+public partial class C
+{
+    void M1(int val)
+    {
+        buffer1[1] = val;
+    }
+
+    int M2()
+    {
+        return buffer2[1];
+    }
+}
+```
+
+will be equivalent to:
+``` C#
+public partial class C
+{
+    void M1(int val)
+    {
+        buffer.AsSpan()[1] = val;
+    }
+
+    int M2()
+    {
+        return buffer2.AsReadOnlySpan()[1];
+    }
+}
+```
+
+#### Metadata import
+
+When compiler imports a field declaration of type *T* and the following conditions are all met:
```

Comment:
If this going to kick even for types in the current compilation unit? If somebody manually field of the right shape manually, is it going to be treated as C# fixed size buffer?

Reply:
> Is this going to kick even for types in the current compilation unit? If somebody manually field of the right shape manually, is it going to be treated as C# fixed size buffer?

Yes. In general, we don't make distinction between types in source and  type in metadata for the purpose of deciding if they match a pattern.